app/src/main/java/ar/com/wolox/android/cookbook/twitterlogin/TwitterLoginRecipePresenter.kt

```diff
@@ -0,0 +1,196 @@
+package ar.com.wolox.android.cookbook.twitterlogin
+
+import android.content.Intent
+import ar.com.wolox.android.cookbook.twitterlogin.adapter.TwitterLoginAdapter
+import ar.com.wolox.android.cookbook.twitterlogin.adapter.TwitterLoginAuthListener
+import ar.com.wolox.android.cookbook.twitterlogin.adapter.TwitterLoginCredentialsListener
+import ar.com.wolox.android.cookbook.twitterlogin.adapter.TwitterLoginEmailListener
+import ar.com.wolox.android.cookbook.twitterlogin.adapter.TwitterLoginPictureListener
+import ar.com.wolox.android.cookbook.twitterlogin.model.YoutubeEmailResponse
+import ar.com.wolox.wolmo.core.presenter.BasePresenter
+import com.twitter.sdk.android.core.TwitterCore
+import com.twitter.sdk.android.core.TwitterSession
+import com.twitter.sdk.android.core.models.User
+import javax.inject.Inject
+
+class TwitterLoginRecipePresenter @Inject constructor(
+    private val twitterAdapter: TwitterLoginAdapter
+) : BasePresenter<TwitterLoginRecipeView>() {
+
+    override fun onViewAttached() {
+        super.onViewAttached()
+
+        initLoginButton()
+        defaultTwitterLogin()
+    }
+
+    /**
+     * Attach callback to TwitterLoginButton is mandatory. When user click btn, if callback
+     * isn't attached, the app throws an exception (null callback)
+     **/
+    private fun initLoginButton() {
+        view.setLoginCallback(twitterAdapter.twitterCallback(object : TwitterLoginAuthListener {
+            override fun onAuthSuccess(result: TwitterSession) {
+                fetchTwitterEmail(result)
+            }
+
+            override fun onAuthError(message: String?) {
+                if (message != null) {
+                    view.showError(message)
+                } else {
+                    view.showAuthFail()
+                }
+            }
+
+            override fun onAuthFail() {
+                view.showAuthFail()
+            }
+        }))
+    }
+
+    fun onDefaultBtnRequest() {
+        if (view.isNetworkAvailable()) {
```

Comment:
I don't like the idea of Interface giving information to Presenter. This is a MVP model. Try to change this.